Fix Vulkan: SpotLight3D's and OmniLight3D's Projector doesn't work

kidinashell opened the pull request:
Fix Vulkan: SpotLight3D's and OmniLight3D's Projector doesn't work (see #59488)
_Edit for updating description after amending commit_

**Important:** The PR shows a change for GLES3 as well to make sure the check for a valid projector texture is done. Please be aware that that doesn't change, that there doesn't exist an implementation for displaying a projector in GLES3 as of yet.

~~I'm not quite sure what the idea behind the previous implementation was, after looking at the `Light` class it doesn't make much sense to me. I looked for the easiest way to check if the property `projector` has any value and since it's of type `RID` as well, which gives us the `is_valid()` method to check if the internal `_id` property is non-ZERO, I went with that.~~

Previous implementation checked for a valid Light RID instead of a Texture RID. Adjusted that to make sure it checks against the correct storage type (see suggestion from Bastiaan below)

Tested on Windows 10 with Vulkan:

- Validation check: OK
- Display of projector: OK

Tested on Windows 10 with GLES3:

- Validation check: OK
- Display of projector: None (wasn't implemented in the first place) 

Prerequisites are still the same: Spotlight3D or OmniLight3D WITH shadows enabled (as per Notification in the Editor).
Info: The texture - should one be assigned - will be projected with shadows disabled as well from certain camera angles, just not correctly.

With projector:
![with_projector](https://user-images.githubusercontent.com/61805849/190922766-ba26fcc4-444c-40a4-a43c-31c872fe8b91.png)

Without projector:
![without-projector](https://user-images.githubusercontent.com/61805849/190922768-8880780c-aa91-4df9-a0ea-35039ea3bf54.png)

*Bugsquad edit:*
- Fixes #59488

kidinashell committed:
Fix Vulkan: SpotLight3D's and OmniLight3D's Projector doesn't work

BastiaanOlij commented:
Ah, it's definitely wrong that it's calling `light_owner.owns` because it's a texture, not the light. Checking `is_valid` is possibly not good enough because you could set it to any RID, not just a proper texture. It should probably be calling `TextureStorage::get_singleton()->owns_texture`, that is probably also where I went wrong because it would have been texture_owner before it got moved into it's own class and I somehow changed it into light_owner having a brainfart :). So if you could try if that works and update the PR?

Also note that the same bug is in the GLES3 renderer so if you could change it there as well that be great.

kidinashell commented:
@BastiaanOlij I can do that for sure. My thinking was just that while setting the projector property it already checks if the given RID is actually a texture or not. (via `texture_add_to_decal_atlas` and then typecasting in the `HashMap` of the texture atlas itself)
Should I go forward with the changes anyway? Thank you for your feedback.

BastiaanOlij commented:
Yeah some of the checking is a little bit dual sometimes but it can't hurt. 

It's in `drivers\gles3\storage\light_storage.h/cpp`

kidinashell commented:
Got you. Getting on it after some sleep :)

kidinashell commented:
@BastiaanOlij 
I fixed the vulkan side of it and that works. I also changed the check in the GLES3 driver, but projectors don't seem to work there. Are projector's even implemented to be used with the GLES3 driver? I mean, it does the same check on https://github.com/godotengine/godot/blob/63c0dc690e06731224e88911ed16d1b798b681b5/servers/rendering/renderer_scene_cull.cpp#L1667, I set a breakpoint after and checked the `light_data` and the projector check is correct (true when there, false when not).
The check works, but it's not showing.

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:
